Provide the watchEditor method as a service

wvanlint opened the pull request:
Provides a service that exposes the `AutocompleteManager` in an asynchronous way.
I'm not convinced on the naming so let me know if you have any suggestions.

nathansobo commented on lib/main.js:
```diff
@@ -44,6 +44,10 @@ module.exports = {
     return this.autocompleteManager
   },
 
+  provideManagerAsync () {
```
Do you think we need to add the suffix `async` to the name of this method? Seems like we don't need it but I don't care *that* much.

wvanlint replied:
I wanted to emphasize that the service returns a function that takes a callback instead of the actual manager. The `async` suffix might be a bit overspecified, so I changed it to `provideAutocompleteManagerCaller`. Let me know if that still sounds off.

nathansobo commented on lib/main.js:
```diff
@@ -44,6 +44,10 @@ module.exports = {
     return this.autocompleteManager
   },
 
+  provideAutocompleteManagerCaller () {
```
See how the `consumeFoo` services are provided below? It seems like something like `provideManager` could blend in well with those. I see where you're coming from wanting the name to document that it's a promise. I feel like that's an issue in a lot of places when returning promises, but I think it really only makes sense to distinguish when there are both sync and async variants of a method. In that case, I think the norm in JS is to decorate the synchronous version with `sync`.

So, to be clear, when the service is consumed it will actually be a promise? It's unfortunate that we didn't design the system to automatically wait on promises before calling consumers in situations like this.

nathansobo commented:
One more thing... I went ahead and merged the documentation that you added in the other PR. Could you add an example in this PR that clarifies the interface? Thanks for this.

wvanlint commented:
I adjusted the naming and added some example code to the documentation. :) 
As for the question, the service will actually be a function that will take a callback on the `AutocompleteManager` instead of a promise.

nathansobo commented on README.md:
````diff
@@ -84,4 +84,30 @@ If the default `SymbolProvider` is missing useful information for the language /
+```
+In the main module file:
+```
+consumeAutocompleteManager(service) {
````
I'm really wondering if we can simplify this service interface. Like could we just expose `autocomplete.watchEditor` as a method you can call to watch an editor. I'm going to spend some time tomorrow thinking about it. If I end up coming up with an idea, I'll just give it a try and get your feedback since you've churned on this a lot.

wvanlint replied:
Oh yeah, that does sound possible if that would be better.
How about something like:

```
provideWatchEditor () {
  return (editor, labels, resultCallback) => {
    this.requireAutocompleteManagerAsync((manager) => {
      let disposable = manager.watchEditor(editor, labels)
      resultCallback(disposable)
    }
  }
}
```

nathansobo replied:
What about the same thing that returns a promise resolving to a disposable, I imagine it working like this...

```js
// In a package that consumes the service

async consumeWatchEditorService (watchEditor) {
  const disposable = await watchEditor(editor, labels)
  // ...
}
```

One thing I'm now wondering is why does the autocomplete manager need to be required asynchronously. Is it a performance thing? Because just having a simple synchronous API seems a lot more sane.

wvanlint replied:
Okay, I can give that a try! The only performance reason that I can imagine is that the autocomplete manager creates the providers which might have to do some indexing.

wvanlint commented:
Updated after #876, now exposing the `watchEditor` method.

nathansobo commented on lib/main.js:
```diff
@@ -22,6 +22,10 @@ module.exports = {
   },
 
+  provideWatchEditor () {
+    return this.autocompleteManager.watchEditor
```
Shouldn't you bind this method?

```js
return this.autocompleteManager.watchEditor.bind(this.autocompleteManager)
```

wvanlint replied:
It was prebound before in `autocomplete-manager.js`: `this.watchEditor = this.watchEditor.bind(this)`. I removed it so it's more clear locally.

nathansobo commented:
Cool. This is looking much clearer now that we dropped the async stuff. A couple comments but we're very close.

wvanlint commented on README.md:
````diff
@@ -82,6 +82,26 @@ Great autocomplete depends on having great autocomplete providers. If there is n
+```
+consumeAutocompleteWatchEditor(watchEditor) {
+  this.autocompleteDisposable = watchEditor(
+      this.editor, ['symbol-provider'])
````
Done. Which style guide do you use? I don't seem to find mentions about wrapping continuations in Javascript Standard Style, I think I was going with the Google style guide.

nathansobo commented:
Sorry I missed that pre-existing `bind`. I'm honestly not the greatest at reviewing diffs. You've been super patient through this whole process and I really appreciate it. This is going to be really cool!
